rskj-core/src/main/java/co/rsk/mine/gas/provider/StableMinGasPriceProvider.java

```diff
@@ -18,41 +18,79 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 
 public abstract class StableMinGasPriceProvider implements MinGasPriceProvider {
     private static final Logger logger = LoggerFactory.getLogger("StableMinGasPrice");
+    private static final int ERR_NUM_OF_FAILURES = 20;
+
     private final MinGasPriceProvider fallBackProvider;
     private final long minStableGasPrice;
-    private Long lastMinGasPrice;
-    private long lastUpdateTimeStamp;
-    private final int refreshRateMillis;
+    private final long refreshRateInMillis;
+    private final AtomicInteger numOfFailures = new AtomicInteger();
+    private final AtomicReference<Future<Long>> priceFuture = new AtomicReference<>();
 
+    private volatile long lastMinGasPrice;
+    private volatile long lastUpdateTimeMillis;
 
-    protected StableMinGasPriceProvider(MinGasPriceProvider fallBackProvider, long minStableGasPrice, int refreshRateInSeconds) {
+    protected StableMinGasPriceProvider(MinGasPriceProvider fallBackProvider, long minStableGasPrice, Duration refreshRate) {
         this.minStableGasPrice = minStableGasPrice;
         this.fallBackProvider = fallBackProvider;
-        this.lastMinGasPrice = 0L;
-        this.refreshRateMillis = refreshRateInSeconds * 1000;
+        this.refreshRateInMillis = refreshRate.toMillis();
     }
 
     protected abstract Optional<Long> getBtcExchangeRate();
 
     @Override
     public long getMinGasPrice() {
-        long currentTime = Instant.now().getEpochSecond();
-        if (currentTime - lastUpdateTimeStamp >= refreshRateMillis) {
-            fetchPrice();
+        return getMinGasPrice(false);
+    }
+
+    @VisibleForTesting
+    public long getMinGasPrice(boolean wait) {
+        long currentTimeMillis = System.currentTimeMillis();
+        if (currentTimeMillis - lastUpdateTimeMillis >= refreshRateInMillis) {
+            Future<Long> priceFuture = fetchPriceAsync();
+            if (wait || priceFuture.isDone()) {
+                try {
+                    return priceFuture.get();
+                } catch (InterruptedException e) {
+                    logger.error("Min gas price fetching was interrupted", e);
+                    Thread.currentThread().interrupt();
+                } catch (ExecutionException e) {
+                    logger.error("Min gas price fetching was failed", e);
+                }
+            }
         }
-        // time it is always updated, it is not taken into account if the price was really updated or not.
-        lastUpdateTimeStamp = currentTime;
-        if (lastMinGasPrice != null && lastMinGasPrice > 0) {
+
+        return getLastMinGasPrice();
+    }
+
+    @Override
+    public Coin getMinGasPriceAsCoin() {
+        return Coin.valueOf(getMinGasPrice());
+    }
+
+    @VisibleForTesting
+    public Coin getMinGasPriceAsCoin(boolean wait) {
```

Comment:
question:

If you already have the annotation `@VisibleForTesting` do you really need of this method to be public? It could be default or protected, no?